module/zcommon/zfs_uio.c

```diff
@@ -98,6 +106,50 @@ uiomove(void *p, size_t n, enum uio_rw rw, struct uio *uio)
 	}
 	return (0);
 }
+
+static int
+uiomove_bvec(void *p, size_t n, enum uio_rw rw, struct uio *uio)
+{
+	const struct bio_vec *bv = uio->uio_bvec;
+	size_t skip = uio->uio_skip;
+	ulong_t cnt;
+
+	ASSERT3U(skip, <, bv->bv_len);
+
+	while (n && uio->uio_resid) {
+		void *paddr;
+		cnt = MIN(bv->bv_len - skip, n);
+
+		paddr = zfs_kmap_atomic(bv->bv_page, KM_USER1);
```

Comment:
You are correct that there is code in the mainline kernel doing this, but the routines doing this likely need to be changed. When CONFIG_PREEMPT is set, the kernel scheduler should be able to preempt us. If the scheduler preempts us and schedules us to another CPU or runs something else on this CPU that touches this, bad things will happen. See the comment for `__schedule`:

http://lxr.free-electrons.com/source/kernel/sched/core.c#L2693

All we need is an IRQ to occur when the scheduler thinks something else should run. As the comment for `__schedule()` indicates, the code in arch/x86/entry/entry_64.S will call into the scheduler on return from the IRQ and `preempt_schedule_irq()` will be called. That will call `schedule()` and something else can have the CPU.